### ISSUE TITLE
Resource-intensive URL cache expiration bogs down Synapse

### ISSUE THREAD

**Issue by spantaleev** [matrix-org/synapse#2638]
The "expire URL cache" feature introduced in #2478 causes Synapse to become unresponsive while the cleanup operation happens.

I've been running a homeserver on an older Synapse version (pre-#2478) and naturally had a long list of entries (thousands) that were due for expiration.

As soon as I upgraded to a newer Synapse version (post-#2478) the cleanup feature kicked in.

It seems that while `PreviewUrlResource._expire_url_cache_data` is doing its job, Synapse is unable not do anything else (handling API requests, federation, etc.)

Since the `media-store` is on a remote filesystem for me (using s3fs-fuse), deleting a file takes a long time.
Doing that for thousands of files one after the other, naturally, takes a very long time.
Synapse cannot do anything else during that time.

I understand that normally such excessive delete operations would not happen. It only happened this time, because cleanup had never happened before (because it wasn't implemented) and there were a lot of things queued up.

Still, people running large homeservers with a media store powered by a slow filesystem (NFS? etc.) would probably also be affected by this to some extent - even during "incremental" cleanup (although doing it once every 10 seconds probably keeps the batches small).

In any case, it seems inappropriate for some maintenance operation to bog down the entire server and prevent it from doing its core responsibility - handling API requests and federation.

**Comment by richvdh** [comment]
yes, it looks like we're doing the FS operations in the main reactor thread rather than delegating them to a separate thread as we should. On a local FS that will be unnoticeable but for a remote FS it will stop synapse responding for significant periods :/

**Comment by cybershaman** [comment]
I want to chime in here, having the same problem. 
Matrix-Synapse spun itself into 10 second intervals of single "expire_url_cache_data" tasks effectively ceasing all normal S2S and C2S operations. This goes on for hours now and while I realise the backing database is huge (which begs to question why it grew so big in the first place) I do not think synapse should cease all operation while running these maintenance tasks.

matrix-synapse version: matrix-synapse 1.12.4-1 (from Debian unstable repo)
database backend: PostgreSQL 9.6
number of local users: <10

Log message:
```
Mai 19 15:33:11 matrix-synapse synapse[1223]: synapse.rest.media.v1.preview_url_resource: [expire_url_cache_data-30] Running url preview cache expi
Mai 19 15:33:11 matrix-synapse synapse[1223]: synapse.rest.media.v1.preview_url_resource: [expire_url_cache_data-30] Deleted 0 media from url cache
Mai 19 15:33:21 matrix-synapse synapse[1223]: synapse.rest.media.v1.preview_url_resource: [expire_url_cache_data-31] Running url preview cache expi
Mai 19 15:33:21 matrix-synapse synapse[1223]: synapse.rest.media.v1.preview_url_resource: [expire_url_cache_data-31] Deleted 0 media from url cache
```

**Comment by richvdh** [comment]
@cybershaman those logs suggest that `expire_url_cache_data` is running quickly. What makes you believe that this is preventing other operations?

**Comment by hex-m** [comment]
For us Synapse stopped working because the NFS-mounted `media_store_path` didn't respond because of a problem on the NFS-Server side. Maybe it have been better to use a `soft` mount for that path? Then it would at least have timed-out.

I didn't create a separate issue because the underlying issue is that core functionality shouldn't be blocked by I/O operations.
